File: lib/presentation/widgets/rating_stars.dart
import 'package:flutter/material.dart';

/// A star rating widget. Read-only by default; pass [onChanged] to make
/// it interactive.
class RatingStars extends StatelessWidget {
  const RatingStars({
    super.key,
    required this.rating,
    this.max = 5,
    this.size = 28,
    this.color = const Color(0xFFFFB300),
    this.onChanged,
  });

  final double rating;
  final int max;
  final double size;
  final Color color;
  final ValueChanged<int>? onChanged;

  @override
  Widget build(BuildContext context) {
    return Row(
      mainAxisSize: MainAxisSize.min,
      children: List.generate(max, (i) {
        final filled = i < rating.floor();
        final half = !filled && i < rating;
        final icon = filled
            ? Icons.star
            : half
                ? Icons.star_half
                : Icons.star_border;
        final star = Icon(icon, size: size, color: color);
        return onChanged == null
            ? star
            : GestureDetector(
                onTap: () => onChanged!(i + 1),
                child: Padding(
                  padding: const EdgeInsets.symmetric(horizontal: 2),
                  child: star,
                ),
              );
      }),
    );
  }
}

File: android_native/app/src/main/java/com/example/widgetgallery/presentation/components/RatingStars.kt
package com.example.widgetgallery.presentation.components

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.size
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Star
import androidx.compose.material.icons.filled.StarBorder
import androidx.compose.material.icons.filled.StarHalf
import androidx.compose.material3.Icon
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import kotlin.math.floor

/** Read-only or interactive star rating. */
@Composable
fun RatingStars(
    rating: Float,
    modifier: Modifier = Modifier,
    max: Int = 5,
    size: Dp = 28.dp,
    color: Color = Color(0xFFFFB300),
    onRatingChange: ((Int) -> Unit)? = null,
) {
    Row(modifier) {
        for (i in 0 until max) {
            val filled = i < floor(rating).toInt()
            val half = !filled && i < rating
            val icon = when {
                filled -> Icons.Filled.Star
                half -> Icons.Filled.StarHalf
                else -> Icons.Filled.StarBorder
            }
            Icon(
                icon, null, tint = color,
                modifier = Modifier
                    .size(size)
                    .then(
                        if (onRatingChange != null)
                            Modifier.clickable { onRatingChange(i + 1) }
                        else Modifier
                    ),
            )
        }
    }
}
